File: 2018/04/rs/src/main.rs
use std::env;
use std::fs;
use std::time::Instant;
use regex::Regex;
use std::collections::HashMap;

pub struct LogRecord {
    pub datetime: (u32, u32, u32, u32, u32),
    pub message: String,
}

pub struct GuardRecord {
    pub total: u32,
    pub minutes: Box<HashMap<u32, u32>>,
}

impl GuardRecord {
    fn new(total: u32) -> GuardRecord {
        GuardRecord {
            total,
            minutes: Box::new((0..60).map(|minute| (minute, 0u32)).collect()),
        }
    }
}

const FALL_ASLEEP: &str = "falls asleep";
const WAKE_UP: &str = "wakes up";
fn build_guard_records(log: Vec<LogRecord>) -> HashMap<u32, GuardRecord> {
    let shift_start_regex: Regex = Regex::new(r"^Guard\s#(?P<id>\d+)\sbegins\sshift").unwrap();
    let mut log = log;
    log.sort_by_key(|a| a.datetime);
    let mut guards: HashMap<u32, GuardRecord> = HashMap::new();
    let mut record_guard_times = |guard_id, last_asleep, woke| {
        if !guards.contains_key(&guard_id) {
            guards.insert(guard_id, GuardRecord::new(0));
        }
        let mut guard_record = guards.get_mut(&guard_id).unwrap();
        for minute in last_asleep..woke {
            *guard_record.minutes.get_mut(&minute).unwrap() += 1;
            guard_record.total += 1;
        }
    };
    let mut guard_id = 0;
    let mut guard_asleep = false;
    let mut last_asleep = 0;
    for log_record in log {
        let log_minutes = log_record.datetime.4;
        if log_record.message == FALL_ASLEEP {
            last_asleep = log_minutes;
            guard_asleep = true;
        } else if log_record.message == WAKE_UP {
            guard_asleep = false;
            record_guard_times(guard_id, last_asleep, log_minutes);
        } else if let Some(capture) = shift_start_regex.captures(&log_record.message) {
            if guard_asleep {
                record_guard_times(guard_id, last_asleep, 60);
                guard_asleep = false;
            }
            guard_id = capture.name("id").unwrap().as_str().parse().unwrap();
        }
    }
    guards
}

fn part1(guards: &HashMap<u32, GuardRecord>) -> u32 {
    let mut max_total = 0;
    let mut guard_id = 0;
    for (id, guard_record) in guards {
        if guard_record.total > max_total {
            max_total = guard_record.total;
            guard_id = *id;
        }
    }
    max_total = 0;
    let mut max_minute = Default::default();
    for (minute, total) in guards.get(&guard_id).unwrap().minutes.iter() {
        if *total > max_total {
            max_total = *total;
            max_minute = *minute;
        }
    }
    guard_id * max_minute
}

fn part2(guards: &HashMap<u32, GuardRecord>) -> u32 {
    let mut max_total = 0;
    let mut guard_id = 0;
    let mut max_minute = 0;
    for (id, guard_record) in guards {
        for (minute, total) in guard_record.minutes.iter() {
            if *total > max_total {
                max_total = *total;
                max_minute = *minute;
                guard_id = *id;
            }
        }
    }
    guard_id * max_minute
}

fn solve(guards: &HashMap<u32, GuardRecord>) -> (u32, u32) {
    (part1(guards), part2(guards))
}

fn parse_line(line: &str) -> LogRecord {
    let line_regex: Regex = Regex::new(r"\[(?P<year>\d{4})-(?P<month>\d{2})-(?P<day>\d{2})\s(?P<hours>\d{2}):(?P<minutes>\d{2})\]\s(?P<message>.*)$").unwrap();
    line_regex
        .captures(line)
        .map(|cap| LogRecord {
            datetime: (
                cap.name("year").unwrap().as_str().parse().unwrap(),
                cap.name("month").unwrap().as_str().parse().unwrap(),
                cap.name("day").unwrap().as_str().parse().unwrap(),
                cap.name("hours").unwrap().as_str().parse().unwrap(),
                cap.name("minutes").unwrap().as_str().parse().unwrap(),
            ),
            message: String::from(cap.name("message").unwrap().as_str()),
        })
        .unwrap()
}

fn get_input(file_path: &String) -> HashMap<u32, GuardRecord> {
    build_guard_records(
        fs::read_to_string(file_path)
            .expect("Error reading input file!")
            .lines()
            .map(|line| parse_line(line))
            .collect(),
    )
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:7}", now.elapsed().as_secs_f32());
}

File: 2018/04/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <regex.h>
#include <string.h>

typedef struct GuardNode
{
    int id, total, minutes[60];
    struct GuardNode *left, *right;
} GuardNode;

typedef struct
{
    int year, month, day, hour, minute;
    char *message;
} LogRecord;

typedef struct
{
    LogRecord *records;
    int count, size;
} Input;
typedef struct
{
    int part1;
    int part2;
} Results;

GuardNode *addToGuardsIfNotPresent(GuardNode *node, int id)
{
    if (node == NULL)
    {
        node = malloc(sizeof(GuardNode));
        node->id = id;
        node->left = node->right = NULL;
        node->total = 0;
        for (int minute = 0; minute < 60; minute++)
            node->minutes[minute] = 0;
    }
    else if (node->id > id)
        node->left = addToGuardsIfNotPresent(node->left, id);
    else if (node->id < id)
        node->right = addToGuardsIfNotPresent(node->right, id);
    return node;
}

GuardNode *getGuardRecord(GuardNode *node, int id)
{
    if (node == NULL)
        return NULL;
    if (node->id == id)
        return node;
    if (node->id > id)
        return getGuardRecord(node->left, id);
    return getGuardRecord(node->right, id);
}

void getMaxTotal(GuardNode *node, int *max, int *id)
{
    if (node == NULL)
        return;
    if (node->total > *max)
    {
        *id = node->id;
        *max = node->total;
    }
    getMaxTotal(node->left, max, id);
    getMaxTotal(node->right, max, id);
}

int part1(GuardNode *guards)
{
    int max = 0, guardId;
    getMaxTotal(guards, &max, &guardId);
    GuardNode *guard = getGuardRecord(guards, guardId);
    int maxTotal = 0, maxMinute;
    for (int minute = 0; minute < 60; minute++)
        if (guard->minutes[minute] > maxTotal)
        {
            maxTotal = guard->minutes[minute];
            maxMinute = minute;
        }
    return guardId * maxMinute;
}

void getMaxMinuteTotal(GuardNode *node, int *max, int *maxMinute, int *id)
{
    if (node == NULL)
        return;
    for (int minute = 0; minute < 60; minute++)
        if (node->minutes[minute] > *max)
        {
            *max = node->minutes[minute];
            *maxMinute = minute;
            *id = node->id;
        }
    getMaxMinuteTotal(node->left, max, maxMinute, id);
    getMaxMinuteTotal(node->right, max, maxMinute, id);
}

int part2(GuardNode *guards)
{
    int maxTotal = 0, maxMinute = -1, guardId;
    getMaxMinuteTotal(guards, &maxTotal, &maxMinute, &guardId);
    return guardId * maxMinute;
}

int compare(int a, int b)
{
    return (a < b) - (b < a);
}

int greaterThan(LogRecord log1, LogRecord log2)
{
    int comp = compare(log1.year, log2.year);
    if (comp == -1)
        return 1;
    if (comp == 1)
        return 0;
    comp = compare(log1.month, log2.month);
    if (comp == -1)
        return 1;
    if (comp == 1)
        return 0;
    comp = compare(log1.day, log2.day);
    if (comp == -1)
        return 1;
    if (comp == 1)
        return 0;
    comp = compare(log1.hour, log2.hour);
    if (comp == -1)
        return 1;
    if (comp == 1)
        return 0;
    return compare(log1.minute, log2.minute) == -1;
}

void sort(Input input)
{
    int i, j;
    LogRecord temp;
    for (i = 0; i < input.count - 1; i++)
        for (j = 0; j < input.count - 1 - i; j++)
            if (greaterThan(input.records[j], input.records[j + 1]))
            {
                temp = input.records[j];
                input.records[j] = input.records[j + 1];
                input.records[j + 1] = temp;
            }
}

void recordGuardTimes(GuardNode *guards, int guardId, int lastAsleep, int woke)
{
    GuardNode *guardRecord = getGuardRecord(guards, guardId);
    for (; lastAsleep < woke; lastAsleep++)
    {
        guardRecord->total++;
        guardRecord->minutes[lastAsleep]++;
    }
}

GuardNode *buildRecords(Input input)
{
    const char *FALL_ASLEEP = "falls asleep", *WAKE_UP = "wakes up";
    sort(input);
    GuardNode *guards = NULL;
    int guardId = 0, newGuardId = 0, guardAsleep = 0, lastAsleep = 0;
    while (input.count--)
    {
        if (strcmp(input.records->message, FALL_ASLEEP) == 0)
        {
            lastAsleep = input.records->minute;
            guardAsleep = 1;
        }
        else if (strcmp(input.records->message, WAKE_UP) == 0)
        {
            guardAsleep = 0;
            recordGuardTimes(guards, guardId, lastAsleep, input.records->minute);
        }
        else if (sscanf(input.records->message, "Guard #%d begins shift", &newGuardId))
        {
            if (guardAsleep)
            {
                recordGuardTimes(guards, guardId, lastAsleep, 60);
                guardAsleep = 0;
            }
            guardId = newGuardId;
            guards = addToGuardsIfNotPresent(guards, guardId);
        }
        input.records++;
    }
    return guards;
}

Results solve(Input input)
{
    GuardNode *guards = buildRecords(input);
    return (Results){part1(guards), part2(guards)};
}

#define INPUT_INCREMENT 10
void addToInput(Input *input, LogRecord record)
{
    if (input->count == input->size)
    {
        input->size += INPUT_INCREMENT;
        input->records = realloc(input->records, input->size * sizeof(LogRecord));
    }
    input->records[input->count++] = record;
}

char *rtrim(char *str, const char *seps)
{
    int i;
    if (seps == NULL)
        seps = "\t\n\v\f\r ";
    i = strlen(str) - 1;
    while (i >= 0 && strchr(seps, str[i]) != NULL)
        str[i--] = '\0';
    return str;
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    regex_t regexCompiled;
    regmatch_t groupArray[7];
    if (regcomp(&regexCompiled, "\\[([0-9]{4})-([0-9]{2})-([0-9]{2}) ([0-9]{2}):([0-9]{2})\\] (.*)$", REG_EXTENDED))
    {
        perror("Error compiling regex.");
        exit(1);
    }
    Input input = {
        calloc(INPUT_INCREMENT, sizeof(LogRecord)),
        0,
        INPUT_INCREMENT};
    char *line = NULL, *cursor = NULL, *message;
    size_t lineLength;
    int group, groupLength, year, month, day, hour, minute;
    while (getline(&line, &lineLength, file) != EOF)
    {
        cursor = line;
        if (!regexec(&regexCompiled, cursor, 7, groupArray, 0))
        {
            for (group = 0; group <= 6 && groupArray[group].rm_so != -1; group++)
            {
                groupLength = strlen(cursor) + 1;
                char cursorCopy[groupLength];
                strcpy(cursorCopy, cursor);
                cursorCopy[groupArray[group].rm_eo] = 0;
                switch (group)
                {
                case 1:
                    year = atoi(cursorCopy + groupArray[group].rm_so);
                    break;
                case 2:
                    month = atoi(cursorCopy + groupArray[group].rm_so);
                    break;
                case 3:
                    day = atoi(cursorCopy + groupArray[group].rm_so);
                    break;
                case 4:
                    hour = atoi(cursorCopy + groupArray[group].rm_so);
                    break;
                case 5:
                    minute = atoi(cursorCopy + groupArray[group].rm_so);
                    break;
                case 6:
                    message = malloc(groupLength);
                    strcpy(message, cursorCopy + groupArray[group].rm_so);
                    rtrim(message, NULL);
                }
            }
            cursor += groupArray[0].rm_eo;
            addToInput(&input, (LogRecord){year, month, day, hour, minute, message});
        }
        else
        {
            perror("Bad format line");
            perror(line);
            exit(1);
        }
    }
    fclose(file);
    return input;
}

void freeInput(Input input)
{
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}
